vowpalwabbit/global_data.cc

```diff
@@ -437,4 +438,45 @@ vw::vw()
   sd->multiclass_log_loss = 0;
   sd->holdout_multiclass_log_loss = 0;
 }
-IGNORE_DEPRECATED_USAGE_END
\ No newline at end of file
+IGNORE_DEPRECATED_USAGE_END
+
+vw::~vw()
+{
+  if (l != nullptr)
+  {
+    l->finish();
+    free(l);
+  }
+
+  // Check if options object lifetime is managed internally.
+  if (should_delete_options)
+    delete options;
+
+  // TODO: migrate all finalization into parser destructor
+  if (p != nullptr)
+  {
+    free_parser(*this);
+    finalize_source(p);
+    p->parse_name.clear();
+    p->parse_name.delete_v();
+    delete p;
+  }
+
+  const bool seeded = weights.seeded() > 0;
+  if (!seeded)
+  {
+    if (sd->ldict)
+    {
+      sd->ldict->~namedlabels();
+      free(sd->ldict);
+    }
+    free(sd);
+  }
+  for (size_t i = 0; i < final_prediction_sink.size(); i++)
+    if (final_prediction_sink[i] != 1)
```

Comment:
nit: What's our bracket style, feel like this is unnecessary might as well change it up in this pr? also for each loop?

Reply:
Kind of depends who you ask. I prefer brackets always though personally. Good point, can definitely be a for each.